src/NLog/Targets/Target.cs

```diff
@@ -722,6 +724,39 @@ protected string RenderLogEvent(Layout layout, LogEventInfo logEvent)
             }
         }
 
+        /// <summary>
+        /// Renders the event info in layout.
+        /// </summary>
+        /// <param name="layout">The layout.</param>
+        /// <param name="logEvent">The event info.</param>
+        /// <param name="defaultValue">The default value when <paramref name="layout"/> and/or <paramref name="logEvent"/> is <c>null</c></param>
+        /// <returns>Value representing log event.</returns>
+        protected T RenderLogEvent<T>([CanBeNull] Layout<T> layout, [CanBeNull] LogEventInfo logEvent, T defaultValue = default(T))
+        {
+            if (layout == null || logEvent == null)
+                return defaultValue;
+
+            if (OptimizeBufferReuse)
+            {
+                if (layout.IsFixed)
+                {
+                    return layout.RenderToValueOrDefault(logEvent, defaultValue);
+                }
+
+                if (TryGetCachedValue(layout, logEvent, out var value))
```

Comment:
This code-path handles layoutrenderer like `${threadid}` when target is wrapped in AsyncWrapperTarget. But `Layout<int>` is not allowed to cast the cached `string` value directly to int-value. Instead this logic should be moved into  `Layout<T>`-class where it can also test if TryGetRawValue can be used with needing help from `ReusableLayoutBuilder`.

Reply:
Is this https://github.com/NLog/NLog/issues/3994?

Reply:
Nope it is bug. Try using `${threadid}` as EventId in the EventLog-target and wrap it in AsyncTargetWrapper. Then it should throw a nice cast-exception

Reply:
Hi @snakefoot,

I think this in the real blocker for this pr (others comments could be resolved without much hassle)

I don't really understand the issue here. Could you please help me with this by writing an unit test or if it's possible, the fix? I really like to merge this PR!

Feel free to push to this branch.

Reply:
It is about correct handling of LayoutRenderers that are not threadagnostic (Like `${threadid}`). And using them with `Layout<int>` in a target that are wrapped with AsyncWrapper.

Reply:
Please help me with an unit test for this. I'm very limited with NLog these days and haven't really time to puzzle to reproduce this.

Reply:
@snakefoot please suggest which unit test and how, or feel free to add an commit to this branch :)

I don't see how I could change the current tests (which is using RenderToValueOrDefault), to the asyncwrapper

Reply:
Fixed with the last commit (this is the same issue, right?)

Thanks, great I understand now the issue and it has been fixed :)